src/services/refactors/moveToFile.ts

```diff
@@ -0,0 +1,1114 @@
+    kind: "refactor.move.file",
+};
+registerRefactor(refactorNameForMoveToFile, {
+    kinds: [moveToFileAction.kind],
+    getAvailableActions: function getRefactorActionsToMoveToFile(context): readonly ApplicableRefactorInfo[] {
+        const statements = getStatementsToMove(context);
+        if (context.preferences.allowTextChangesInNewFiles && statements) {
+            return [{ name: refactorNameForMoveToFile, description, actions: [moveToFileAction] }];
+        }
+        if (context.preferences.provideRefactorNotApplicableReason) {
+            return [{ name: refactorNameForMoveToFile, description, actions:
+                [{ ...moveToFileAction, notApplicableReason: getLocaleSpecificMessage(Diagnostics.Selection_is_not_a_valid_statement_or_statements) }]
+            }];
+        }
+        return emptyArray;
+    },
+    getEditsForAction: function getRefactorEditsToMoveToFile(context, actionName, newFile): RefactorEditInfo | undefined {
+        Debug.assert(actionName === refactorNameForMoveToFile, "Wrong refactor invoked");
+        const statements = Debug.checkDefined(getStatementsToMove(context));
+        Debug.assert(newFile, "Target file does not exist");
+        const edits = textChanges.ChangeTracker.with(context, t => doChange(context, context.file, newFile, context.program, statements, t, context.host, context.preferences));
+        return { edits, renameFilename: undefined, renameLocation: undefined };
+    }
+});
+
+function doChange(context: RefactorContext, oldFile: SourceFile, newFile: string, program: Program, toMove: ToMove, changes: textChanges.ChangeTracker, host: LanguageServiceHost, preferences: UserPreferences): void {
+    const checker = program.getTypeChecker();
+    const usage = getUsageInfo(oldFile, toMove.all, checker);
+    //creating a new file
+    if (!host.fileExists(newFile) || host.fileExists(newFile) && program.getSourceFile(newFile)?.statements.length === 0) {
+        changes.createNewFile(oldFile, newFile, getNewStatementsAndRemoveFromOldFile(oldFile, newFile, usage, changes, toMove, program, host, newFile, preferences, /*newFileExists*/ false));
+        addNewFileToTsconfig(program, changes, oldFile.fileName, newFile, hostGetCanonicalFileName(host));
+    }
+    else {
+        const sourceFile = program.getSourceFile(newFile);
+        if (sourceFile) {
+            const importAdder = codefix.createImportAdder(sourceFile, context.program, context.preferences, context.host);
+            getNewStatementsAndRemoveFromOldFile(oldFile, newFile, usage, changes, toMove, program, host, newFile, preferences, /*newFileExists*/ true, importAdder);
+        }
+    }
+}
+
+function getNewStatementsAndRemoveFromOldFile(
+    oldFile: SourceFile, newFile: string, usage: UsageInfo, changes: textChanges.ChangeTracker, toMove: ToMove, program: Program, host: LanguageServiceHost, newFilename: string, preferences: UserPreferences, newFileExists: boolean, importAdder?: codefix.ImportAdder

```

Comment:
Clarity nits:

1. Throughout this file, can we do a review of everywhere the word “new” is used and make sure it’s only getting used when the file is actually _new_? I propose to settle on the term “target file” to mean the file to which we’re moving code, whether it is _new_ or _existing_. Everything here seems to be using the terminology “new” as a legacy from when this code was only used for a refactor called “Move to new file.”
2. This function seems to have a duplicated parameter, `newFile` and `newFilename`.
3. You could simplify the signature even more by removing `newFileExists` and making the type of `targetFile` be `string | SourceFile`. If it’s a string, it’s the file name for a new file. If it’s a SourceFile, it’s the existing target file.